libc/time/utimes.c

```diff
@@ -0,0 +1,528 @@
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+
+	/* ENAMETOOLONG */
+	char tooLongPath[PATH_MAX];
+	memset(tooLongPath, 'a', PATH_MAX - 1);
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, utimes(tooLongPath, NULL));
+	TEST_ASSERT_EQUAL_INT(ENAMETOOLONG, errno);
+
+	/* ELOOP */
+	TEST_ASSERT_EQUAL_INT(0, symlink("loop_symlink", "loop_symlink"));
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, utimes("loop_symlink", NULL));
+	TEST_ASSERT_EQUAL_INT(ELOOP, errno);
+	unlink("loop_symlink");
+
+	/* ENOTDIR */
+	int fd;
+	fd = creat("not_a_directory", 0777);
+	close(fd);
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, utimes("not_a_directory/file.txt", NULL));
+	TEST_ASSERT_EQUAL_INT(ENOTDIR, errno);
+	remove("not_a_directory");
+}
+
+
+TEST_GROUP(test_futimes);
+
+
+TEST_SETUP(test_futimes)
+{
+	int fd;
+	TEST_ASSERT_GREATER_THAN_INT(0, fd = creat(FILENAME, 0644));
+	close(fd);
+}
+
+
+TEST_TEAR_DOWN(test_futimes)
+{
+	unlink(FILENAME);
+}
+
+
+TEST(test_futimes, set_null)
+{
+	unsigned int i;
+
+	for (i = 0; i < LOOP_CNT; ++i) {
+		usleep(random() % 500000);
+		test_utimes(FILENAME, TEST_SET_NULL, FUTIMES);
+	}
+}
+
+
+TEST(test_futimes, set_now)
+{
+	unsigned int i;
+
+	for (i = 0; i < LOOP_CNT; ++i) {
+		usleep(random() % 500000);
+		test_utimes(FILENAME, TEST_SET_NOW, FUTIMES);
+	}
+}
+
+
+TEST(test_futimes, set_past)
+{
+	unsigned int i;
+
+	for (i = 0; i < LOOP_CNT; ++i) {
+		test_utimes(FILENAME, TEST_SET_PAST, FUTIMES);
+	}
+}
+
+
+TEST(test_futimes, one)
+{
+	int fd;
+	struct timeval tv[2];
+	struct stat statbuf;
+
+	tv[0].tv_sec = 1;
+	tv[1].tv_sec = 1;
+	tv[0].tv_usec = 1;
+	tv[1].tv_usec = 1;
+
+	TEST_ASSERT_GREATER_THAN_INT(0, fd = open(FILENAME, O_RDONLY));
+	TEST_ASSERT_EQUAL_INT(0, futimes(fd, tv));
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &statbuf));
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	TEST_ASSERT_EQUAL_INT64(tv[0].tv_sec, statbuf.st_atim.tv_sec);
+	TEST_ASSERT_EQUAL_INT64(tv[1].tv_sec, statbuf.st_mtim.tv_sec);
+
+/* FIXME: only second accuracy is supported */
+#ifndef __phoenix__
+	TEST_ASSERT_EQUAL_INT64(tv[0].tv_usec * 1000, statbuf.st_atim.tv_nsec);
+	TEST_ASSERT_EQUAL_INT64(tv[1].tv_usec * 1000, statbuf.st_mtim.tv_nsec);
+#endif
+}
+
+
+TEST(test_futimes, high_value)
+{
+	int fd;
+	struct timeval tv[2];
+	struct stat statbuf;
+
+	tv[0].tv_sec = INT32_MAX;
+	tv[1].tv_sec = INT32_MAX;
+	tv[0].tv_usec = 999999;
+	tv[1].tv_usec = 999999;
+
+	TEST_ASSERT_GREATER_THAN_INT(0, fd = open(FILENAME, O_RDONLY));
+	TEST_ASSERT_EQUAL_INT(0, futimes(fd, tv));
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &statbuf));
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	TEST_ASSERT_EQUAL_INT64(tv[0].tv_sec, statbuf.st_atim.tv_sec);
+	TEST_ASSERT_EQUAL_INT64(tv[1].tv_sec, statbuf.st_mtim.tv_sec);
+
+/* FIXME: only second accuracy is supported */
+#ifndef __phoenix__
+	TEST_ASSERT_EQUAL_INT64(tv[0].tv_usec * 1000, statbuf.st_atim.tv_nsec);
+	TEST_ASSERT_EQUAL_INT64(tv[1].tv_usec * 1000, statbuf.st_mtim.tv_nsec);
+#endif
+}
+
+
+TEST(test_futimes, errnos)
```

Comment:
Please leave a comment why other errnos was not possible to check, or at least leave todo - applies also to utimes/lutimes.